panda/src/windisplay/winGraphicsWindow.cxx

```diff
@@ -3072,39 +3074,54 @@ get_icon(const Filename &filename) {
   }
 
   // If it wasn't found, resolve the filename and search for that.
-
-  // Since we have to use a Windows call to load the image from a filename, we
-  // can't load a virtual file and we can't use the virtual file system.
   Filename resolved = filename;
-  if (!resolved.resolve_filename(get_model_path())) {
-    // The filename doesn't exist along the search path.
-    if (resolved.is_fully_qualified() && resolved.exists()) {
-      // But it does exist locally, so accept it.
+  if ((resolved.resolve_filename(get_model_path())) || (resolved.is_fully_qualified() && resolved.exists())) {
+    // The filename exists in the search path or locally, accepted.
 
-    } else {
+    fi = _icon_filenames.find(resolved);
+    if (fi != _icon_filenames.end()) {
+      _icon_filenames[filename] = (*fi).second;
+      return (HICON)((*fi).second);
+    }
+
+    Filename os = resolved.to_os_specific();
+
+    HANDLE h = LoadImage(nullptr, os.c_str(),
+                        IMAGE_ICON, 0, 0, LR_LOADFROMFILE | LR_DEFAULTSIZE);
+    if (h == 0) {
       windisplay_cat.warning()
-        << "Could not find icon filename " << filename << "\n";
+        << "windows icon filename '" << os << "' could not be loaded!!\n";
       return 0;
     }
-  }
-  fi = _icon_filenames.find(resolved);
-  if (fi != _icon_filenames.end()) {
-    _icon_filenames[filename] = (*fi).second;
-    return (HICON)((*fi).second);
-  }
 
-  Filename os = resolved.to_os_specific();
-
-  HANDLE h = LoadImage(nullptr, os.c_str(),
-                       IMAGE_ICON, 0, 0, LR_LOADFROMFILE | LR_DEFAULTSIZE);
-  if (h == 0) {
+    _icon_filenames[filename] = h;
+    _icon_filenames[resolved] = h;
+    return (HICON)h;
+  } else {
     windisplay_cat.warning()
-      << "windows icon filename '" << os << "' could not be loaded!!\n";
+      << "Could not find icon filename " << filename << " locally, checking VFS.\n";
+  }
+  VirtualFileSystem *vfs = VirtualFileSystem::get_global_ptr();
+  PT(VirtualFile) vfile = vfs->get_file(filename);
+  if (vfile != nullptr) {
+    std::string vfile_data = vfile->read_file(0);
+    long fsize = vfile_data.length();
+
+    int offset = LookupIconIdFromDirectoryEx((PBYTE)vfile_data.data(), TRUE, 0, 0, LR_DEFAULTCOLOR);
+    if (offset != 0) {
+      HICON hIcon = CreateIconFromResourceEx((PBYTE)vfile_data.data() + offset, fsize - offset, TRUE, 0x30000, 0, 0, LR_DEFAULTCOLOR);
```

Comment:
I'm wondering if there is a lifetime issue here. I haven't read the documentation, but does the icon created by CreateIconFromResourceEx copy the data passed in, or does it continue to refer to the same data in memory? If so, that might be an issue since the `std::string` will be deallocated at the end of the function.

Secondly, you're passing in the remaining size of the whole data structure, not the size of the icon in it; is that okay?